### ISSUE TITLE
gRPC doesn't build on manylinux1 (sched_getcpu not found)

### ISSUE THREAD

**Issue by JonasVautherin** [grpc/grpc#19291]
I am building gRPC v1.21.0 from source, on different platforms. When trying with manylinux1 (from dockcross), I get the following error:

```
[  2%] Building CXX object CMakeFiles/grpc_plugin_support.dir/src/compiler/csharp_generator.cc.o
/work/build/manylinux1-x64_CFLAGS2/grpc/grpc/src/grpc/src/core/lib/gpr/cpu_linux.cc: In function ‘void init_num_cpus()’:
/work/build/manylinux1-x64_CFLAGS2/grpc/grpc/src/grpc/src/core/lib/gpr/cpu_linux.cc:40:20: error: ‘sched_getcpu’ was not declared in this scope
   if (sched_getcpu() < 0) {
                    ^
/work/build/manylinux1-x64_CFLAGS2/grpc/grpc/src/grpc/src/core/lib/gpr/cpu_linux.cc: In function ‘unsigned int gpr_cpu_current_cpu()’:
/work/build/manylinux1-x64_CFLAGS2/grpc/grpc/src/grpc/src/core/lib/gpr/cpu_linux.cc:69:26: error: ‘sched_getcpu’ was not declared in this scope
   int cpu = sched_getcpu();
                          ^
```

In the past, I fixed that by forcing the definition of `GPR_MUSL_LIBC_COMPAT` (as can be seen [here](https://github.com/dronecore/grpc/commit/d2126e12dbe510fa3c36442f9a34417ab7b88d0f)). But that is not a proper fix.

I have seen that `tools/run_tests/artifacts/artifact_targets.py` sets manylinux from the CFLAGS:

```
environ['CFLAGS'] = '-DGPR_MANYLINUX1=1'
```

But I don't know how that is related to my musl issue above, and setting `CFLAGS` did not seem to solve this issue.

I created a small repo showing how I build everything [here](https://github.com/JonasVautherin/grpc-build). Here is a way to reproduce the issue:

1. Build gRPC with `cmake -Bbuild/default -S.` to convince yourself that it works.
2. Get dockcross for manylinux1-x64:
```
    $ docker run --rm dockcross/manylinux1-x64 > ./dockcross-manylinux1-x64
```
3. Try to build with dockcross and see the error described above:
```
    $ ./dockcross-manylinux1-x64 cmake -Bbuild/manylinux1-x64 -S.
```
4. Try with `CFLAGS` set and see that the error stays:
```
    $ ./dockcross-manylinux1-x64 bash -c 'export CFLAGS="-DGPR_MANYLINUX1=1"; cmake -Bbuild/manylinux1-x64 -S.'
```

@gnossen: Would you have an advice about that? Am I missing something?

**Comment by gnossen** [comment]
It looks like your flag isn't getting plumbed through. [We define `GPR_CPU_POSIX` on manylinux](https://github.com/grpc/grpc/blob/bb6a97b0c8fc11a5c27b03e2058dc383bbfc1e64/include/grpc/impl/codegen/port_platform.h#L106). This file mentioned in your error should basically be filtered out by the preprocessor if the flag gets plumbed through correctly.

Try adding the flags to the `CMAKE_FLAGS` variable used in the grpc external project. Like so
```cmake
list(APPEND CMAKE_ARGS
    "-DCMAKE_INSTALL_PREFIX:PATH=${CMAKE_INSTALL_PREFIX}"
    ...
    "-DCMAKE_CXX_FLAGS=-DGPR_MANYLINUX1=1"
    )
```
That will get you past your undefined symbol. But it looks like there are still some other problems with the build: 

```
_gRPC_CPP_PLUGIN-NOTFOUND: program not found or is not executable
--grpc_out: protoc-gen-grpc: Plugin failed with status code 1.
gmake[5]: *** [gens/src/proto/grpc/status/status.grpc.pb.cc] Error 1
gmake[4]: *** [CMakeFiles/grpc++_error_details.dir/all] Error 2
gmake[4]: *** Waiting for unfinished jobs....
_gRPC_CPP_PLUGIN-NOTFOUND: program not found or is not executable
_gRPC_CPP_PLUGIN-NOTFOUND: program not found or is not executable
--grpc_out: protoc-gen-grpc: Plugin failed with status code 1.
--grpc_out: protoc-gen-grpc: Plugin failed with status code 1.
```

CC @nicolasnoble

**Comment by JonasVautherin** [comment]
Right, so I was not setting the flag properly, thanks for that! And I also get the grpc_cpp_plugin issue, but I don't get it because the file is there in:

```
./build/manylinux1-x64/grpc/grpc/src/grpc-build/grpc_cpp_plugin
```

**Comment by JonasVautherin** [comment]
Oh wait, I believe that's related to the other fix I was doing in the older version (see [here](https://github.com/grpc/grpc/commit/4a7e0c653bf5aa629b5fa28fd11228ce5b017ec9)). Let me try that.

**Comment by JonasVautherin** [comment]
Yes, I believe that's it: [this branch](https://github.com/JonasVautherin/grpc-build/tree/manylinux) builds successfully with:

```
docker run --rm dockcross/manylinux1-x64 > ./dockcross-manylinux1-x64
./dockcross-manylinux1-x64 cmake -Bbuild/manylinux1-x64 -S.
```

I think that's related to a ["known peculiarity"](https://cmake.org/cmake/help/latest/variable/CMAKE_CROSSCOMPILING.html) (as in: I cannot imagine a scenario where anybody would want CMake to behave like this, but it is documented :sweat_smile:):

> [CMAKE_CROSSCOMPILING] will be set to true by CMake if the CMAKE_SYSTEM_NAME variable has been set manually (i.e. in a toolchain file or as a cache entry from the cmake command line). In most cases, manually setting CMAKE_SYSTEM_NAME will only be done when cross compiling, since it will otherwise be given the same value as CMAKE_HOST_SYSTEM_NAME if not manually set, which is correct for the non-cross-compiling case. In the event that CMAKE_SYSTEM_NAME is manually set to the same value as CMAKE_HOST_SYSTEM_NAME, then CMAKE_CROSSCOMPILING will still be set to true.

So my workaround is to point to [a fork of gRPC](https://github.com/JonasVautherin/grpc-build/blob/manylinux/grpc/CMakeLists.txt#L42), which [applies this patch](https://github.com/dronecore/grpc/commit/6c7702e3cfc119c3dcc0d6370ccb7059df9cb581):

```
     #if cross-compiling, find host plugin
-    if(CMAKE_CROSSCOMPILING)
+    if(CMAKE_CROSSCOMPILING AND NOT CMAKE_SYSTEM_NAME STREQUAL CMAKE_HOST_SYSTEM_NAME)
        find_program(_gRPC_CPP_PLUGIN grpc_cpp_plugin)
     else()
        set(_gRPC_CPP_PLUGIN $<TARGET_FILE:grpc_cpp_plugin>)
```

i.e. `CMAKE_CROSSCOMPILING AND NOT CMAKE_SYSTEM_NAME STREQUAL CMAKE_HOST_SYSTEM_NAME`.

@nicolasnoble @gnossen: Do you have an opinion about that? I would see two ways to patch it:

1. The way I did it above, but I don't like it because for every `if(CMAKE_CROSSCOMPILING)`, one will have to wonder if it applies to manylinux or not.
2. At the top of the root CMakeLists.txt, "fix" `CMAKE_CROSSCOMPILING` (i.e. disable it if `CMAKE_SYSTEM_NAME STREQUAL CMAKE_HOST_SYSTEM_NAME`). The disadvantage of this is that it is changing the official cmake behavior, but it sounds like a documented bug to me anyway.

**Comment by JonasVautherin** [comment]
I'm thinking that this could be solved in a more elegant way from the manylinux toolchain, actually. I [opened an issue](https://github.com/dockcross/dockcross/issues/326) in the dockcross repo.

Regarding `-DGPR_MANYLINUX1`, I see the following lines [in your build](https://github.com/grpc/grpc/blob/bb6a97b0c8fc11a5c27b03e2058dc383bbfc1e64/include/grpc/impl/codegen/port_platform.h#L101-L103):

```c++
#elif defined(GPR_MANYLINUX1)
// TODO(atash): manylinux1 is just another __linux__ but with ancient
// libraries; it should be integrated with the `__linux__` definitions below.
```

I am not sure exactly what it would imply to "integrate with the `__linux__` definitions. Do you think it is feasible to remove this `GPR_MANYLINUX1` variable? Or is it somehow required to know that it is a manylinux build, in which case the manylinux toolchain should probably provide a `MANYLINUX1` variable instead of you setting `GPR_MANYLINUX1`?

**Comment by gnossen** [comment]
I can't say I fully understand what the TODO here means either. Unfortunately, the author has since left the team.

If we were to get rid of `GPR_MANYLINUX1`, what would the alternative be in your mind? I see two approaches to configuring a native build. 

@nicolasnoble Please feel free to correct/disagree on anything I'm about to say.

The first is prescriptive. The user tells the compiler exactly which libc and system features to expect. This is what we're doing with `GPR_MANYLINUX1`. We *do* use this in our official `manylinux` wheel build. I don't particularly have a problem with this, because it allows us to use the standard `setuptools` facilities to compile our C extension.

The second is exploratory. An initial probing step is executed to determine which features are available on the platform on which the artifact is being built. CMake, Autoconf, and configure scripts are prime examples of this type. Since you're using Cmake, I would have expected this to "just work". I think the problem here with `sched_getcpu` is it was introduced in glibc 2.6, but we have no preprocessor version check on it.

**TL;DR** I think we should:
1. Leave `GPR_MANYLINUX1` as it is.
2. Make platform detection smarter for `sched_getcpu` and any other assumptions that fall apart on older version of glibc.

**Comment by JonasVautherin** [comment]
> 1.    Leave GPR_MANYLINUX1 as it is.
> 2.    Make platform detection smarter for sched_getcpu and any other assumptions that fall apart on older version of glibc.

Well a third way would be to set `MANYLINUX1` in the manylinux toolchain directly, so you can do `if (MANYLINUX1)`, just like you do `if (IOS)`, `if (ANDROID)` or `if (MINGW)`. To me, the question is the following:

"Is manylinux just an old linux, or does it need to be handled differently?"

And in my understanding, it could actually be considered like another platform, just like Android: manylinux has some limitations because e.g. of the old glibc, so it is legit for a build system to do some `if (MANYLINUX1)` in some cases. In which case I would try to set `MANYLINUX1` [directly at the manylinux toolchain level](https://github.com/dockcross/dockcross/issues/327). Does that make sense to you?

**Comment by gnossen** [comment]
In my opinion, [`manylinux1` legitimately is "just an old Linux"](https://www.python.org/dev/peps/pep-0513/). It's based off of Centos 5.11 with a very limited set of shared object libraries installed, all with *old* ABIs.

With that said, it certainly is a valid choice to treat it as its own platform. In fact, it's very similar to what we're doing right now. But I'm not holding out hope that the manylinux maintainers will be responsive to the idea.

**Comment by JonasVautherin** [comment]
Good news: both PRs got merged into dockcross, meaning that in the next release, we will have:

* `MANYLINUX1` or `MANYLINUX2010` defined in the toolchain (see [here](https://github.com/dockcross/dockcross/pull/330))
* `CMAKE_CROSSCOMPILING` will work as we expect, i.e. will be FALSE for manylinux (see [here](https://github.com/dockcross/dockcross/pull/329)).

No action needed then, but it will allow us to remove `GPR_MANYLINUX1` in favor of `MANYLINUX1`, and it will automatically fix the issue I was experiencing with `CMAKE_CROSSCOMPILING` :+1:.

Closing this issue, and I'll try to open a PR once `MANYLINUX1` can be used.

**Comment by gnossen** [comment]
@JonasVautherin Amazing! Thank you!
